discordLevelingSystem/rank_card.py

```diff
@@ -0,0 +1,222 @@
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+DEALINGS IN THE SOFTWARE.
+"""
+
+from io import BufferedIOBase, BytesIO, IOBase
+from os import PathLike
+from typing import Optional, Union
+
+from aiohttp import ClientSession
+from PIL import Image, ImageDraw, ImageFont
+from .errors import InvalidImageType, InvalidImageUrl
+from .member_data import MemberData
+
+class RankCard:
+    """Represents the rank card that will be sent to the member upon leveling up
+
+    Parameters
+    ----------
+    background: :class:`Union[PathLike, BufferedIOBase]`
+        The background image for the rank card. This can be a path to a file or a file-like object in `rb` mode
+
+    avatar: :class:`Union[PathLike, BufferedIOBase]`
+        The avatar image for the rank card. This can be a path to a file or a file-like object in `rb` mode
+    
+    level: :class:`int`
+        The level of the member
+    
+    username: :class:`str`
+        The username of the member
+    
+    current_exp: :class:`int`
+        The current amount of XP the member has
+    
+    max_exp: :class:`int`
+        The amount of XP required for the member to level up
+    
+    bar_color: :class:`Optional[str]`
+        The color of the XP bar. This can be a hex code or a color name. Default is `white`
+    
+    text_color: :class:`Optional[str]`
+        The color of the text. This can be a hex code or a color name. Default is `white`
+    
+    path: :class:`Optional[PathLike]`
+        The path to save the rank card to. If this is not provided, `bytes` will be returned instead.
+    
+    member_data: :class:`Optional[MemberData]`
+        The member data to use for the rank card. If this is provided, the other parameters will be ignored.
+
+    Attributes
+    ----------
+    - `background`
+    - `avatar`
+    - `level`
+    - `username`
+    - `current_exp`
+    - `max_exp`
+    - `bar_color`
+    - `text_color`
+    - `path`
+    - `member_data`
+    """
+
+    __slots__ = ('background', 'avatar', 'level', 'username', 'current_exp', 'max_exp', 'bar_color', 'text_color', 'path', 'member_data')
+
+
+
+    def __init__(
+        self,
+        background: Union[PathLike, BufferedIOBase],
+        avatar: Union[PathLike, BufferedIOBase],
+        level:int,
+        username:str,
+        current_exp:int,
+        max_exp:int,
+        bar_color:Optional[str]="white",
+        text_color:Optional[str]="white",
+        path:Optional[str]=None,
+        member_data:Optional[MemberData]=None
+    )-> None:
+        self.background = background
+        self.avatar = avatar
+        if member_data is not None:
+            self.level = member_data.level
+            self.username = member_data.name
+            self.current_exp = member_data.xp
+            self.max_exp = member_data.total_xp
+
+        else:
+            self.level = level
+            self.username = username
+            self.current_exp = current_exp
+            self.max_exp = max_exp
```

Comment:
I feel this would be better removed and handled internally through a `get_rank_card()` method. See my main comment.

Reply:
your `get_data_for()` not being static makes it a lot harder for me to implement this
either i have to add it to the main `DiscordLevelingSystem` or make the user give it as a parameter

Reply:
Right. Adding it to the main leveling system is what I had in mind.

**leveling_system.py**
```py
async def get_rank_card(self, member: Member) -> io.BytesIO :
        data = await self.get_data_for(member)
        from . import RankCard # this wouldnt be here but its just a quick example
        card = RankCard(
            background=open('./background.png', mode='rb'),
            avatar=open('./av_rdr.jpg', mode='rb'), # needs to be changed to a URL but just quick example
            level=data.level,
            username=data.name,
            current_exp=data.xp,
            max_exp=self.get_xp_for_level(await self.next_level(member))
        )
        return await card.card()
```
**End user code**
```py
@bot.command()
async def test(ctx: commands.Context):
    card_data = await lvl.get_rank_card(ctx.author)
    await ctx.send(file=discord.File(card_data, filename='rank.png'))
```
**Results in**

![rank](https://user-images.githubusercontent.com/52070034/190718133-e88db67d-d1d2-403f-a896-f5c499cbfb7d.png)